# Task 1

Import libraries

In [ ]:
import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
# needed to display chart in Jupyter notebooks
%matplotlib inline 

import sklearn
import scipy
import scipy.stats as stats

In [2]:
adm_df = pd.read_csv('Admissions.csv')
lab_df = pd.read_csv('Lab.csv')
trans_df = pd.read_csv('Transfusions.csv')

Let's take a look at each df

In [3]:
adm_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5231 entries, 0 to 5230
Data columns (total 11 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   admission_id                5231 non-null   int64  
 1   patient_id                  5231 non-null   int64  
 2   admission_date              5231 non-null   object 
 3   admission_time              4535 non-null   object 
 4   discharge_date              5231 non-null   object 
 5   discharge_time              5231 non-null   object 
 6   hospital                    5231 non-null   object 
 7   age                         4604 non-null   float64
 8   sex                         5231 non-null   object 
 9   charlson_comorbidity_index  4970 non-null   object 
 10  lap_score                   5231 non-null   int64  
dtypes: float64(1), int64(3), object(7)
memory usage: 449.7+ KB


adm_df contains 5231 individual records in total. 
Lots of missing values for admission_time, age, and charlson_comorbidity_index
Also, admission and discharge date and time are not properly formatted, since they are objects, not datetime 

In [4]:
adm_df.head()

admission_id  patient_id admission_date admission_time discharge_date  \
0          3602       34208     2011-09-07            NaN     09/11/2011   
1          2790       90930     2008-07-22          06:19     07/24/2008   
2          8755       97765     2002-08-24          14:28     08/27/2002   
3          9260       17941     2007-01-12          09:39     01/26/2007   
4          5317       36532     2003-09-24          13:37     10/12/2003   

  discharge_time                           hospital   age sex  \
0          13:16         St. Joseph's Health Centre   NaN   M   
1          17:54             St. Michael's Hospital  48.0   M   
2          08:14               Mount Sinai Hospital  59.0   F   
3          11:10  Sunnybrook Health Sciences Centre  57.0   M   
4          16:23           Toronto Western Hospital  51.0   F   

  charlson_comorbidity_index  lap_score  
0                          1         45  
1                          1         46  
2                          0         35  
3                          1         18  
4                          1         21

In [5]:
adm_df.describe()

admission_id    patient_id         age    lap_score
count   5231.000000   5231.000000  4604.00000  5231.000000
mean    5512.069776  54170.202829    54.94874    27.528197
std     2584.436372  26121.399502     5.05146    21.423248
min     1003.000000  10017.000000    38.00000     0.000000
25%     3280.500000  30970.000000    51.00000    11.000000
50%     5501.000000  54736.000000    55.00000    23.000000
75%     7760.500000  76768.500000    58.00000    40.000000
max     9997.000000  99984.000000    74.00000   136.000000

Here are some the descriptive stats on the continuous variables from adm_df

Observations:
    admission_id and patient_id seems to be just a random unique number, not in ascending order
    age has missing values, but it's mostly people in their 50s
    lap_score has quite a large standard deviation (almost as much as the mean)
    admission and discharge date format is not consistent, and in some cases the admission time is missing (this needs to be taken care of in the data cleaning process)
    

In [6]:
lab_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3609 entries, 0 to 3608
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   admission_id  3609 non-null   int64  
 1   test_name     3609 non-null   object 
 2   test_code     3609 non-null   object 
 3   result_unit   3609 non-null   object 
 4   result_value  3609 non-null   float64
 5   result_date   3609 non-null   object 
 6   result_time   3609 non-null   object 
dtypes: float64(1), int64(1), object(5)
memory usage: 197.5+ KB


In [7]:
lab_df.describe()

admission_id  result_value
count   3609.000000   3609.000000
mean    5451.744250     81.539762
std     2609.854333     94.470176
min     1003.000000      0.000000
25%     3194.000000      3.600000
50%     5413.000000     48.600000
75%     7715.000000    140.900000
max     9997.000000    584.400000

In [8]:
lab_df.head()

admission_id        test_name test_code result_unit  result_value  \
0          4416          Albumin       ALB        g/dL           4.1   
1          4636   Platelet Count       PLT        E9/L         295.0   
2          8041       Hemoglobin       RBC         g/L         165.4   
3          7384  Chloride plasma      CLPL      mmol/L         193.0   
4          2604          Albumin       ALB        g/dL           6.0   

  result_date result_time  
0  2003-08-19       02:25  
1  2012-02-05       01:42  
2  2004-04-18       08:05  
3  2000-04-04       14:55  
4  2008-02-01       11:30

Observations:


In [9]:
trans_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 657 entries, 0 to 656
Data columns (total 6 columns):
 #   Column                Non-Null Count  Dtype 
---  ------                --------------  ----- 
 0   admission_id          657 non-null    int64 
 1   issue_date            657 non-null    object
 2   issue_time            657 non-null    object
 3   rbc_transfusion       657 non-null    bool  
 4   platelet_transfusion  657 non-null    bool  
 5   plasma_transfusion    657 non-null    bool  
dtypes: bool(3), int64(1), object(2)
memory usage: 17.4+ KB


In [10]:
trans_df.describe()

admission_id
count    657.000000
mean    5561.634703
std     2577.215753
min     1010.000000
25%     3408.000000
50%     5619.000000
75%     7760.000000
max     9997.000000

In [11]:
trans_df.head()

admission_id  issue_date issue_time  rbc_transfusion  platelet_transfusion  \
0          8755  2002-08-25      16:05            False                 False   
1          9736  2006-03-17      08:05             True                 False   
2          4152  2002-07-18      19:08             True                 False   
3          7077  2010-10-06      08:07             True                 False   
4          9297  2010-01-22      21:11             True                 False   

   plasma_transfusion  
0                True  
1               False  
2               False  
3               False  
4               False

Observations:


# Task 2

In [12]:
all_df = adm_df.merge(lab_df, on = 'admission_id', how='left')
all_df = all_df.merge(trans_df, on = 'admission_id', how='left')

In [13]:
all_df.head()

admission_id  patient_id admission_date admission_time discharge_date  \
0          3602       34208     2011-09-07            NaN     09/11/2011   
1          2790       90930     2008-07-22          06:19     07/24/2008   
2          8755       97765     2002-08-24          14:28     08/27/2002   
3          9260       17941     2007-01-12          09:39     01/26/2007   
4          5317       36532     2003-09-24          13:37     10/12/2003   

  discharge_time                           hospital   age sex  \
0          13:16         St. Joseph's Health Centre   NaN   M   
1          17:54             St. Michael's Hospital  48.0   M   
2          08:14               Mount Sinai Hospital  59.0   F   
3          11:10  Sunnybrook Health Sciences Centre  57.0   M   
4          16:23           Toronto Western Hospital  51.0   F   

  charlson_comorbidity_index  ...  test_code result_unit result_value  \
0                          1  ...        NaN         NaN          NaN   
1                          1  ...        ALB        g/dL          2.2   
2                          0  ...        ALB        g/dL          1.8   
3                          1  ...        RBC         g/L        172.1   
4                          1  ...        ALB        g/dL          2.5   

  result_date  result_time  issue_date issue_time rbc_transfusion  \
0         NaN          NaN         NaN        NaN             NaN   
1  2008-07-22        16:45         NaN        NaN             NaN   
2  2002-08-26        15:27  2002-08-25      16:05           False   
3  2007-01-12        20:10         NaN        NaN             NaN   
4  2003-09-27        19:31         NaN        NaN             NaN   

  platelet_transfusion plasma_transfusion  
0                  NaN                NaN  
1                  NaN                NaN  
2                False               True  
3                  NaN                NaN  
4                  NaN                NaN  

[5 rows x 22 columns]

In [14]:
all_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 5231 entries, 0 to 5230
Data columns (total 22 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   admission_id                5231 non-null   int64  
 1   patient_id                  5231 non-null   int64  
 2   admission_date              5231 non-null   object 
 3   admission_time              4535 non-null   object 
 4   discharge_date              5231 non-null   object 
 5   discharge_time              5231 non-null   object 
 6   hospital                    5231 non-null   object 
 7   age                         4604 non-null   float64
 8   sex                         5231 non-null   object 
 9   charlson_comorbidity_index  4970 non-null   object 
 10  lap_score                   5231 non-null   int64  
 11  test_name                   3609 non-null   object 
 12  test_code                   3609 non-null   object 
 13  result_unit                 3609 

The merges seemed to work, so now let's calculate the percentage of missing values for each column:

In [15]:
all_df.isnull().sum()

admission_id                     0
patient_id                       0
admission_date                   0
admission_time                 696
discharge_date                   0
discharge_time                   0
hospital                         0
age                            627
sex                              0
charlson_comorbidity_index     261
lap_score                        0
test_name                     1622
test_code                     1622
result_unit                   1622
result_value                  1622
result_date                   1622
result_time                   1622
issue_date                    4574
issue_time                    4574
rbc_transfusion               4574
platelet_transfusion          4574
plasma_transfusion            4574
dtype: int64

In [16]:
percent_missing = all_df.isnull().sum() * 100 / len(all_df)
missing_value_df = pd.DataFrame({'percent_missing': percent_missing})

In [17]:
missing_value_df

percent_missing
admission_id                       0.000000
patient_id                         0.000000
admission_date                     0.000000
admission_time                    13.305295
discharge_date                     0.000000
discharge_time                     0.000000
hospital                           0.000000
age                               11.986236
sex                                0.000000
charlson_comorbidity_index         4.989486
lap_score                          0.000000
test_name                         31.007456
test_code                         31.007456
result_unit                       31.007456
result_value                      31.007456
result_date                       31.007456
result_time                       31.007456
issue_date                        87.440260
issue_time                        87.440260
rbc_transfusion                   87.440260
platelet_transfusion              87.440260
plasma_transfusion                87.440260

It worked. Seems like Issue_date and time, rbc_transfusion, platelet and plasma transfusion are missing most of their values. Good to know.

Let's move on to the next question.

# Task 3

Let's fill in the NAs of admission time with the most popular value

In [18]:
# filling in the NAs
adm_df['admission_time'] = adm_df['admission_time'].fillna(adm_df['admission_time'].value_counts().index[0])

# Now merge again
all_df = adm_df.merge(lab_df, on = 'admission_id', how='left')
all_df = all_df.merge(trans_df, on = 'admission_id', how='left')


In [19]:
all_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 5231 entries, 0 to 5230
Data columns (total 22 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   admission_id                5231 non-null   int64  
 1   patient_id                  5231 non-null   int64  
 2   admission_date              5231 non-null   object 
 3   admission_time              5231 non-null   object 
 4   discharge_date              5231 non-null   object 
 5   discharge_time              5231 non-null   object 
 6   hospital                    5231 non-null   object 
 7   age                         4604 non-null   float64
 8   sex                         5231 non-null   object 
 9   charlson_comorbidity_index  4970 non-null   object 
 10  lap_score                   5231 non-null   int64  
 11  test_name                   3609 non-null   object 
 12  test_code                   3609 non-null   object 
 13  result_unit                 3609 

Now let's convert all the admission and discharge dates and times to datetime objects

In [20]:
all_df.apply(lambda x: x['admission_date'] +" "+ x['admission_time'], axis=1)

0       2011-09-07 21:58
1       2008-07-22 06:19
2       2002-08-24 14:28
3       2007-01-12 09:39
4       2003-09-24 13:37
              ...       
5226    2002-12-27 16:49
5227    2008-01-26 12:58
5228    2007-05-16 15:10
5229    2007-02-19 09:50
5230    2010-02-23 21:37
Length: 5231, dtype: object

In [21]:
admission_datetime = pd.to_datetime(all_df.apply(lambda x: x['admission_date'] +" "+ x['admission_time'], axis=1))
discharge_datetime = pd.to_datetime(all_df.apply(lambda x: x['discharge_date'] +" "+ x['discharge_time'], axis=1))

Conversion looks good, so now let's subtract the dates, then get the mean length_of_stay at each hospital

In [22]:
length_of_stay = discharge_datetime - admission_datetime

In [23]:
length_of_stay

0       3 days 15:18:00
1       2 days 11:35:00
2       2 days 17:46:00
3      14 days 01:31:00
4      18 days 02:46:00
             ...       
5226   15 days 05:56:00
5227    1 days 12:47:00
5228    3 days 15:54:00
5229   11 days 17:17:00
5230    6 days 06:14:00
Length: 5231, dtype: timedelta64[ns]

In [24]:
all_df['length_of_stay'] = length_of_stay.values.astype(np.int64)

In [25]:
pd.to_timedelta(all_df.groupby('hospital').mean()['length_of_stay'])

hospital
Mount Sinai Hospital                11 days 05:52:47.176781002
St. Joseph's Health Centre          10 days 08:09:52.931034482
St. Michael's Hospital              10 days 17:47:57.154989384
Sunnybrook Health Sciences Centre   11 days 00:07:46.744730679
Toronto Western Hospital            11 days 03:35:16.897506925
Name: length_of_stay, dtype: timedelta64[ns]

It worked. Looks like the means of all hospitals are around 10 or 11 days.

# Task 4

Let's fill in the NA values of age with the mean age of the sampled population. I can not think out anything quicker that does not rely on a linear model to predict the age. Inputting the most popular age would not help either, since it would be even more biased. This will affect the statistical model significanly, since the age values will be pulled more towards the mean, introducing bias.

In [26]:
all_df['age'].fillna(all_df['age'].mean(), inplace=True)

In [27]:
all_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 5231 entries, 0 to 5230
Data columns (total 23 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   admission_id                5231 non-null   int64  
 1   patient_id                  5231 non-null   int64  
 2   admission_date              5231 non-null   object 
 3   admission_time              5231 non-null   object 
 4   discharge_date              5231 non-null   object 
 5   discharge_time              5231 non-null   object 
 6   hospital                    5231 non-null   object 
 7   age                         5231 non-null   float64
 8   sex                         5231 non-null   object 
 9   charlson_comorbidity_index  4970 non-null   object 
 10  lap_score                   5231 non-null   int64  
 11  test_name                   3609 non-null   object 
 12  test_code                   3609 non-null   object 
 13  result_unit                 3609 

For the charlson-comorbidity-index I quickly found 1 website: https://www.mdcalc.com/charlson-comorbidity-index-cci

saying that:
 - for <50 years charlson-comorbidity-index = 0
 - for 50–59 years charlson-comorbidity-index = +1
 - for 60–69 years charlson-comorbidity-index = +2
 - for 70–79 years charlson-comorbidity-index = +3
 - for ≥80 years charlson-comorbidity-index = +4

In [28]:
def calculate_charlson_comorbidity_index(age):
    if age < 50:
        return "0"
    elif age < 60:
        return "1+"
    elif age < 70:
        return "2+"
    elif age < 79:
        return "3+"
    else:
        return "4+"

In [29]:
all_df['charlson_comorbidity_index'].fillna(lambda x: calculate_charlson_comorbidity_index(x['age']), axis=0, inplace=True)

In [30]:
all_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 5231 entries, 0 to 5230
Data columns (total 23 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   admission_id                5231 non-null   int64  
 1   patient_id                  5231 non-null   int64  
 2   admission_date              5231 non-null   object 
 3   admission_time              5231 non-null   object 
 4   discharge_date              5231 non-null   object 
 5   discharge_time              5231 non-null   object 
 6   hospital                    5231 non-null   object 
 7   age                         5231 non-null   float64
 8   sex                         5231 non-null   object 
 9   charlson_comorbidity_index  5231 non-null   object 
 10  lap_score                   5231 non-null   int64  
 11  test_name                   3609 non-null   object 
 12  test_code                   3609 non-null   object 
 13  result_unit                 3609 

Seems like it worked, and all the age, and charlson_comorbidity_index missing values have been imputated. Some assumptions have been made, but overall it's better to have some assumed values instead of removing the whole sample from the data set due to missing values.

# Task 5

# Task 6

# Task 7

In [31]:
from sklearn.linear_model import LinearRegression

In [38]:
linear_regressor = LinearRegression()

test_results_df = all_df[all_df['result_value'].isnull()==False]

In [40]:
test_results_df['sex'] = test_results_df['sex'].astype('category')
test_results_df['sex'] = test_results_df['sex'].cat.codes
print(test_results_df['sex'])

1       1
2       0
3       1
4       0
7       0
       ..
5219    0
5222    1
5228    1
5229    1
5230    1
Name: sex, Length: 3609, dtype: int8


c:\users\norbert\appdata\local\programs\python\python36\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.
c:\users\norbert\appdata\local\programs\python\python36\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


In [ ]:


all_df['hospital'] = all_df['hospital'].astype('category')
all_df['hospital'] = all_df['hospital'].cat.codes
print(all_df['hospital'])

all_df.info()

In [36]:
dependent = all_df[['age', 'sex', 'hospital']]
independent = all_df['result_value']
linear_regressor.fit(dependent, independent)
Y_pred = linear_regressor.predict(dependent)

ValueError: Input contains NaN, infinity or a value too large for dtype('float64').

In [ ]:
# The coefficients
print('Coefficients: \n', regr.coef_)
# The mean squared error
print('Mean squared error: %.2f'
      % mean_squared_error(diabetes_y_test, diabetes_y_pred))
# The coefficient of determination: 1 is perfect prediction
print('Coefficient of determination: %.2f'
      % r2_score(diabetes_y_test, diabetes_y_pred))

# Plot outputs
plt.scatter(diabetes_X_test, diabetes_y_test,  color='black')
plt.plot(diabetes_X_test, diabetes_y_pred, color='blue', linewidth=3)

plt.xticks(())
plt.yticks(())

plt.show()

# Task 8

# Task 9

# Task 10